Sources/SwiftGRPCNIO/HTTP1ToRawGRPCServerCodec.swift

```diff
@@ -27,111 +27,200 @@ public enum RawGRPCServerResponsePart {
+          ctx.fireChannelRead(self.wrapInboundOut(.message(buffer)))
+        } else {
+          ctx.fireChannelRead(self.wrapInboundOut(.message(slice)))
+        }
+
+        self.buffer = nil
+        bodyState = .expectingCompressedFlag
       }
-      ctx.fireChannelRead(self.wrapInboundOut(.end))
     }
   }
+
+  private func processEnd(ctx: ChannelHandlerContext, trailers: HTTPHeaders?) throws -> InboundState {
+    guard trailers == nil else {
+      throw StateMachineError("unexpected trailers received \(String(describing: trailers))")
+    }
+
+    ctx.fireChannelRead(self.wrapInboundOut(.end))
+    return .ignore
+  }
 }
 
 extension HTTP1ToRawGRPCServerCodec: ChannelOutboundHandler {
   public typealias OutboundIn = RawGRPCServerResponsePart
   public typealias OutboundOut = HTTPServerResponsePart
-  
+
   public func write(ctx: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+    if case .ignore = outboundState { return }
+
     let responsePart = self.unwrapOutboundIn(data)
     switch responsePart {
     case .headers(let headers):
+      guard case .expectingHeaders = outboundState else { return }
+
       //! FIXME: Should return a different version if we want to support pPRC.
       ctx.write(self.wrapOutboundOut(.head(HTTPResponseHead(version: .init(major: 2, minor: 0), status: .ok, headers: headers))), promise: promise)
+      outboundState = .expectingBodyOrStatus
+
     case .message(var messageBytes):
-      // Write out a length-delimited message payload. See `channelRead` fpor the corresponding format.
-      var responseBuffer = ctx.channel.allocator.buffer(capacity: messageBytes.readableBytes + 5)
+      guard case .expectingBodyOrStatus = outboundState else { return }
+
+      // Write out a length-delimited message payload. See `processBodyState` for the corresponding format.
+      var responseBuffer = ctx.channel.allocator.buffer(capacity: messageBytes.readableBytes + protobufMetadataSize)
       responseBuffer.write(integer: Int8(0))  // Compression flag: no compression
       responseBuffer.write(integer: UInt32(messageBytes.readableBytes))
       responseBuffer.write(buffer: &messageBytes)
       ctx.write(self.wrapOutboundOut(.body(.byteBuffer(responseBuffer))), promise: promise)
+      outboundState = .expectingBodyOrStatus
+
     case .status(let status):
       var trailers = status.trailingMetadata
       trailers.add(name: "grpc-status", value: String(describing: status.code.rawValue))
       trailers.add(name: "grpc-message", value: status.message)
-      ctx.write(self.wrapOutboundOut(.end(trailers)), promise: promise)
+
+      // "Trailers-Only" response
+      if case .expectingHeaders = outboundState {
+        trailers.add(name: "content-type", value: "application/grpc")
+        let responseHead = HTTPResponseHead(version: .init(major: 2, minor: 0), status: .ok)
+        ctx.write(self.wrapOutboundOut(.head(responseHead)), promise: nil)
```

Comment:
This will send two header frames — one header with content type and "OK", and a trailer one for the trailers. Should we merge the two instead? Is that even possible, or does HTTP2 always expect a header and a trailer frame?

Reply:
That's correct -- it's not possible in NIO HTTP2, unfortunately. 

The "Trailers-Only" comment is misleading though: the spec says that "Trailers-Only" should be sent in a single frame:

> Response-Headers & Trailers-Only are each delivered in a single HTTP2 HEADERS frame block. Most responses are expected to have both headers and trailers but Trailers-Only is permitted for calls that produce an immediate error.

This is really response headers _and_ trailers. I'll fixup the comment.